Drift chamber digi

BrieucF opened the pull request:
BEGINRELEASENOTES
- Add a data type for Drift Chamber digitized hits (before the left right ambiguity is resolved)

ENDRELEASENOTES

Tagging @atolosadelgado

andresailer commented on edm4hep.yaml:
```diff
@@ -503,6 +503,22 @@ datatypes:
+      - float time [ns] // time of the hit
+      - float eDep [GeV] // energy deposited on the hit
+      - float eDepError [GeV] // error measured on eDep
+      - edm4hep::Vector3d position [mm] // point on the sensitive wire (SW) which is closest to the simhit
```
This reads like it is using Monte Carlo information?

andresailer commented on edm4hep.yaml:
```diff
@@ -503,6 +503,22 @@ datatypes:
+      - float eDep [GeV] // energy deposited on the hit
+      - float eDepError [GeV] // error measured on eDep
+      - edm4hep::Vector3d position [mm] // point on the sensitive wire (SW) which is closest to the simhit
+      - edm4hep::Vector3d directionSW // direction of SW
```
```suggestion
      - edm4hep::Vector3d directionSW // direction of the sensitive wire (SW)
```
? Along the sensitive wire? Whose direction?

SW is like always software, so using SW should be explicit everywhere and never assume people will know this is sensitive wire.

BrieucF replied:
SW spelled in the comment (left `directionSW` as is) and tried to clarify the meaning of this member

andresailer commented on edm4hep.yaml:
```diff
@@ -503,6 +503,22 @@ datatypes:
+      - edm4hep::Vector3d position [mm] // point on the sensitive wire (SW) which is closest to the simhit
+      - edm4hep::Vector3d directionSW // direction of SW
+      - float distanceToWire [mm] // distance hit-wire
+      - uint32_t nCluster // number of clusters associated to this hit
```
Is this really needed if the vector member exists anyway?
Add a helper function to return nCluster from `nElectrons.size()`?

BrieucF replied:
Done

andresailer commented on edm4hep.yaml:
```diff
@@ -809,6 +825,33 @@ datatypes:
+     - edm4hep::SimTrackerHit to          // reference to the SimTrackerHit
+    ExtraCode:
+      declaration: "
+      [[deprecated(\"use getFrom instead\")]] edm4hep::DriftChamberDigi getRec() const;\n
```
Why are there already deprecated functions added to a new datatype?

BrieucF replied:
They were there in case people wanted to use links in the "Association" way. But maybe we should not let that happen since there is no existing code using this data type. Removed these helpers.

andresailer commented on edm4hep.yaml:
```diff
@@ -503,6 +503,22 @@ datatypes:
       void setCovMatrix(float value, edm4hep::Cartesian dimI, edm4hep::Cartesian dimJ) { getCovMatrix().setValue(value, dimI, dimJ); }\n
       "
 
+  edm4hep::DriftChamberDigi:
```
SensitiveWireHit or DriftChamberHit? 

"DriftChamberDigi" sounds like the algorithm that does the digitisation.

BrieucF replied:
Went for SensitiveWireHit

BrieucF commented:
Thanks Andre

tmadlener commented on edm4hep.yaml:
```diff
@@ -503,6 +503,27 @@ datatypes:
+      - edm4hep::Vector3d directionSW // direction of the sensitive wire (normal vector to the circle)
+      - float distanceToWire [mm] // distance hit-wire (radius of the circle)
+    VectorMembers:
+      - uint16_t nElectrons // number of electrons for each cluster (number of clusters = vector size)
```
What is the typical number of electrons per cluster? I suppose 65535 is enough to work in all cases?

BrieucF replied:
Yes, `uint16` will do the job, this value is typically below 10.

atolosadelgado replied:
> What is the typical number of electrons per cluster? I suppose 65535 is enough to work in all cases?

According to the paper they published [1], Fig 15, right plot:
![image](https://github.com/user-attachments/assets/f47938bd-52d2-4e58-b50a-746897c7a527)
There is a tail that goes up to 450, but the probability is extremely low, most of the clusters have less than 10 electrons as Brieuc said.

[1] https://arxiv.org/pdf/2105.07064

tmadlener replied:
450 still fits into 9 bits, so there is a 7 bit safety margin. I think this is OK. Thanks for the reference.

tmadlener commented on edm4hep.yaml:
```diff
@@ -875,3 +905,4 @@ interfaces:
     Types:
       - edm4hep::TrackerHit3D
       - edm4hep::TrackerHitPlane
+      - edm4hep::SensitiveWireHit
```
How confusing will the slightly different meanings of `position` be in this case? Resp. do we need to worry about this?

BrieucF replied:
Yes, if used blindly together with the position of the other data types in this interface, things will go wrong (a 3D hit position is ill-defined for this new data type). We went for the less intrusive option w.r.t. existing edm4hep but we can remove the position from the interface and name it differently for the `SensitiveWireHit`, if there is a consensus that this is the way to go. Probably a good thing to discuss tomorrow?

tmadlener replied:
I would have a preference for keeping it at the moment. However, I think we should discuss this again to see how to best make this work or if this is truly a concern that actually matters in reality.

tmadlener commented on edm4hep.yaml:
```diff
@@ -503,6 +503,27 @@ datatypes:
+    ExtraCode:
+      declaration: "
+        /// Return the number of clusters associated to the hit\n
+        uint32_t nClusters() const { return getNElectrons().size(); }\n
```
Is there a specific reason to make this a `uint32_t` instead of just going with the `size_t` (or even just `auto`) that `vector::size` returns?

BrieucF replied:
Went for auto

tmadlener commented on edm4hep.yaml:
```diff
@@ -503,6 +503,27 @@ datatypes:
+    ExtraCode:
+      declaration: "
+        /// Return the number of clusters associated to the hit\n
+        uint32_t nClusters() const { return getNElectrons().size(); }\n
```
```suggestion
        uint32_t getNClusters() const { return getNElectrons().size(); }\n
```
To fit with the overall naming schema of getters in EDM4hep.

BrieucF replied:
Done

tmadlener commented on edm4hep.yaml:
```diff
@@ -511,11 +511,13 @@ datatypes:
+      - float eDepError [GeV] // error on eDep
       - edm4hep::Vector3d directionSW // direction of the sensitive wire (normal vector to the circle)
-      - float distanceToWire [mm] // distance hit-wire (radius of the circle)
+      - edm4hep::Vector3d position [mm] // point on the sensitive wire which is closest to the hit (center of the circle)
+      - double positionAlongWireError [mm] // error on the hit position along the wire direction
```
How would the interpretation of the `positionAlongWireError` work in this case? IIUC I would need the `directionSW` to get the direction and then the `position` to get a point on the wire to get a reference point(?).

BrieucF replied:
Correct, the reference point on the wire is `position`, then to know "in which direction the error is defined" without accessing the geometry, you need `sensitiveWireDirection` (which I just renamed for more clarity)

BrieucF commented:
I have added a proposal for the errors. I also applied @gaede suggestion and made the direction vector 2D (which I also renamed to be more clear). Comments welcome

BrieucF commented:
Actually, this data type would be more robust with explicit fields for the wire stereo and azimuthal angle (current proposal updated to this). With the vector2f, users might have different conventions.

gaede commented:
As discussed this morning, you could also make this class look very similar to the `TrackerHitPlane`, e.g. by replacing 
```yaml
- edm4hep::Vector2d sensitiveWireDirection // direction of the sensitive wire (normal vector to the circle)
```
with
```yaml
  - edm4hep::Vector2f u            // measurement direction along the wire - (theta, phi) in global coordinates 
  - edm4hep::Vector2f v            // measurement direction orthogonal to the wire -(theta, phi) in global coordinates 
  - float du                       // measurement error along the wire 
  - float dv                       // measurement error of distance to the wire 
```
this way you'd have the uncertainties in the class, and we would express similar concepts with similar/same names.
Once tracking is done, you could updated `v` such that `m = pos + v` is the actual position of the hit base on the track fit...

tmadlener commented:
Are the `u` and `v` also labelled as *global coordinates* in the `TrackerHitPlane`? I though they are to be interpreted as local coordinates?

gaede commented:
The directions `u` and `v` are meant to be in global coordinates - see: [https://ilcsoft.desy.de/LCIO/current/doc/doxygen_api/html/classEVENT_1_1TrackerHitPlane.html](https://ilcsoft.desy.de/LCIO/current/doc/doxygen_api/html/classEVENT_1_1TrackerHitPlane.html).  There seems sth. has been lost in translation when copying this over from LCIO. Should update our yaml file...

jmcarcell commented:
For the inclusion, if this makes it easier for people to use and test it we could add it to EDM4hep and then we can add warnings to make it clear that it is an experimental type (breaking changes will happen), and also not include it in our "official" CI files. Then, no one should be upset if the files that have collections with these are not readable after it changes in EDM4hep. We should foresee this happening after EDM4hep 1.0 and always denying adding a new type until it is completed is quite restrictive. In addition, if most (all?) of the users of the new type are here and are aware it may change then it should be easier. But I can also see a type being added as an experiment forever...

BrieucF commented:
> > Once tracking is done, you could updated `v` such that `m = pos + v` is the actual position of the hit base on the track fit...
> 
> No such thing as _updating after ..._ in EDM4hep. That would mean cloning + setting a new value in that clone.
> 
> I think if things are conceptually different enough then it makes sense to also keep their abstractions different. Choosing the wrong abstraction might do more damage than not abstracting enough. Since this type should represent the digitized measurement that comes out of a drift chamber, maybe we have to make sure it does that first and only as a (somewhat) secondary concern think about how it relates to tracking, resp. how it would be used in the final fitted track as a tracker hit.
> 
> With that in mind and also regarding some parts of the discussion in #382, I think we will probably have to think a bit about what a _TrackerHit_ actually is, and also what a "raw tracker hit" could be and how we can connect the two things in the end.
> 
> > not include it in our "official" CI files. Then, no one should be upset if the files that have collections with these are not readable after it changes in EDM4hep.
> 
> I think we should still write them, but potentially not check them on the reading side. The reason I say we should still write them is that we don't want to break the rest of the files simply because they contain collections of datatypes that are not yet stable. Currently podio always reads full frames with no way of just skipping certain collections (until we merge [AIDASoft/podio#504](https://github.com/AIDASoft/podio/pull/504)). In principle the buffer creation mechanism takes into account the schema version so that at least connecting the buffers to the branches should work correctly even if things _would_ break during actual schema evolution later (in case of unsupported evolutions).
> 
> Here we have to sit down and define some sort of policy on how we want to deal with this.

This is interesting and we should indeed have this discussion but it goes somewhat beyond the scope of this pull request. I am afraid we won't converge in the timescale within which this PR is needed. Shall I add 'EXPERIMENTAL' to the `Description` field for now?

jmcarcell commented on edm4hep.yaml:
```diff
@@ -486,6 +486,30 @@ datatypes:
+      declaration: "
+        /// Return the number of clusters associated to the hit\n
+        auto getNClusters() const { return getNElectrons().size(); }\n
+        "
```
```suggestion
      declaration: |
        /// Return the number of clusters associated to the hit
        auto getNClusters() const { return getNElectrons().size(); }
```
